gateway-server/src/main/java/org/apache/knox/gateway/services/registry/impl/DefaultServiceDefinitionRegistry.java

```diff
@@ -82,15 +127,152 @@ public void stop() throws ServiceLifecycleException {
+      }
+      return null;
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  @Override
+  public Set<ServiceDefinitionPair> getServiceDefinitions() {
+    readLock.lock();
     try {
-      match = entries.match(Parser.parseLiteral(urlPattern));
-    } catch ( URISyntaxException e ) {
-      LOG.failedToParsePath(urlPattern, e);
+      return serviceDefinitions;
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  @Override
+  public void saveServiceDefinition(ServiceDefinitionPair serviceDefinition) throws ServiceDefinitionRegistryException {
+    saveOrUpdateServiceDefinition(serviceDefinition, false);
+  }
+
+  @Override
+  public void saveOrUpdateServiceDefinition(ServiceDefinitionPair serviceDefinition) throws ServiceDefinitionRegistryException {
+    saveOrUpdateServiceDefinition(serviceDefinition, true);
+  }
+
+  public void saveOrUpdateServiceDefinition(ServiceDefinitionPair serviceDefinition, boolean allowUpdate) throws ServiceDefinitionRegistryException {
+    final ServiceDefinition service = serviceDefinition.getService();
+    final Optional<ServiceDefinition> persistedServiceDefinition = findServiceDefinition(service.getName(), service.getRole(), service.getVersion());
+    if (persistedServiceDefinition.isPresent() && !allowUpdate) {
+      throw new ServiceDefinitionRegistryException("The requested service definition (" + serviceDefinition.toString() + ") already exists!");
+    } else {
+      writeLock.lock();
+      try {
+        final Path serviceDefinitionFolderPath = createServiceDefinitionFolders(service.getName(), service.getVersion());
+        writeOutServiceDefinitionFile(service, serviceDefinitionFolderPath);
+        if (serviceDefinition.getRewriteRules() != null) {
+          writeOutRewriteRules(serviceDefinition.getRewriteRules(), serviceDefinitionFolderPath);
+        }
+        populateServiceDefinitions();
+      } catch (JAXBException | IOException e) {
+        throw new ServiceDefinitionRegistryException("Error while persisting service definition " + serviceDefinition.toString(), e);
+      } finally {
+        writeLock.unlock();
+      }
+      notifyListeners(service.getName(), service.getRole(), service.getVersion());
+    }
+  }
+
+  private Path createServiceDefinitionFolders(String name, String version) throws IOException {
+    final Path serviceDefinitionPath = Paths.get(gatewayConfig.getGatewayServicesDir(), name, version);
+    return Files.createDirectories(serviceDefinitionPath);
+  }
+
+  private void writeOutServiceDefinitionFile(ServiceDefinition serviceDefinition, Path serviceDefinitionFolderPath) throws JAXBException, IOException {
+    final Marshaller marshaller = getJaxbContext().createMarshaller();
+    marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, Boolean.TRUE);
+    try (BufferedWriter serviceDefinitionXmlFileWriter = newBufferedWriter(Paths.get(serviceDefinitionFolderPath.toAbsolutePath().toString() + "/" + SERVICE_DEFINITION_FILE_NAME),
```

Comment:
I think this is creating a path? would be good to use like Paths.get(parent, child) or like .resolve on the Path object. There are a few examples of this.

Reply:
I'll make this change and submit a new patch